utils/create-yaml.js

```diff
@@ -46,187 +22,224 @@ const NEVER_ADD = [
-    } else {
-      out += ` #${ type }`;
-    }
+/**
+ * Looks for collection type annotations and comments out the collection.
+ *
+ * This is neccessary because the library YAML doesn't allow you to create
+ * a comment node and adding comments to the neighboring nodes makes the code
+ * clunky compared to doing a resolve pass.
+ *
+ * Source                   ->            Result
+ * collection:                            #collection:
+ *   #typeAnnotation                        key: value
+ *   key: value
+ * @param {*} documentArray
+ */
+export function resolveCollectionTypeAnnotations(documentArray) {
+  return documentArray.reduce((agg, line) => {
+    const lastLine = (agg.slice(-1)[0] || '');
+    const isTypeAnnotation = line.trim() === `#${ TYPE_ANNOTATION }`;
+    const beggining = isTypeAnnotation ? agg.slice(0, -1) : agg;
+    const newLine = isTypeAnnotation ? commentLine(lastLine) : line;
+
+    return [...beggining, newLine];
+  }, []);
+}
 
-    return out;
-  }
+/**
+ * Looks for scalar type annotations and comments out the scalar.
+ *
+ * Source                   ->            Result
+ * key: '#typeAnnotation - string'        #key: string
+ * @param {*} documentArray
+ */
+export function resolveScalarTypeAnnotations(documentArray) {
+  const annotation = `#${ TYPE_ANNOTATION } - `;
+
+  return documentArray.map((line) => {
+    return line.includes(annotation)
+      ? exports.commentLine(line.replace(annotation, '').replace(/\"+/g, ''))
+      : line;
+  });
 }
 
-function comment(lines) {
-  return (lines || '').split('\n').map(x => `#${ x.replace(/#/g, '') }`).join('\n');
+/**
+ * Looks for collection and scalar type annotations and appropriately comments
+ * out the sections.
+ *
+ * This is neccessary because the library YAML doesn't allow you to create
+ * a comment node and adding comments to the neighboring nodes makes the code
+ * clunky compared to doing a resolve pass.
+ *
+ * @param {YAML.Document} document YAML.Document
+ */
+export function resolveTypeAnnotationsAndConvertToString(document) {
+  const documentAsString = document.toString();
+  let documentAsArray = documentAsString.split('\n');
+
+  documentAsArray = exports.resolveCollectionTypeAnnotations(documentAsArray);
+  documentAsArray = exports.resolveScalarTypeAnnotations(documentAsArray);
+
+  return documentAsArray.join('\n');
 }
 
-function indent(lines, depth = 1) {
-  return _indent(lines, depth * INDENT, ' ', /^#/);
+/**
+ * Given a js object and an expanded schema we will add fields
+ * that are in the schema that are missing from data, remove fields that
+ * are a part of the NEVER_ADD list, add type annotations as comments
+ * and serialize to yaml.
+ * @param {*} data Arbitrary js object
+ * @param {*} expandedSchema A schema with the extra fields expandedResourceFields and expandedSubTypes
+ *                           which links to schemas wrather than typeNames.
+ */
+export function createYaml(data, expandedSchema) {
```

Comment:
This might improve some situation where the old one produces invalid output, but it's now _always_ producing an invalid structure, and is producing unhelpful output that doesn't highlight the parts you need to fill out anymore.

- apiVersion & kind aren't filled out anymore, or at the top of the document
- Things that were in the "ALWAYS_ADD" list are no longer uncommented (e.g. metadata.name).  This is make the common required parts stand out, and be expanded (if it was auto folded).
- Seemingly random things that weren't in ALWAYS_ADD _are_ uncommented, with no obvious reason why.  It looks like every field that is an array of strings?  Lots of random things `metadata.finalizers`, anything with a list of `namespaces` or `values`. 
- If a field is uncommented, all of its parent keys should be too.  The output of this is just syntactically invalid because there's keys floating in space at random indentations.  It should always produce something valid, even if you're missing fields that are going to be "required" when submitted.:
```yaml
#spec:
# selector:
#   matchExpressions:
#     - key: string
#       operator: string
        values:
```

Reply:
Can you list the requirements or what the intention is? I thought the intention was to add commented out sections for the entire schema so that it's easier for the user to fill out their yaml.

The uncommented sections are definitely a bug so I'll fix that and I'll implement the requirements once you list them.

Reply:
There's a lot of back and forth with me and Darren baked into how the folding works, which things are uncommented, what style of YAML it is (i.e. extra indent before dashes in arrays), etc. So as a starting point for testing, the current output and behavior of the editor is effectively the desired behavior (± bugs like however you get `[object Object]`).

The basic ideas are:
  - Construct the entire tree so that you can see all possible fields
  - Hide some fields that you're very unlikely to want (mostly in metadata)
  - Comment out fields that are optional so they're ignored unless edited
  - Leave things that are required or very important as uncommented keys, so that the parts you _must_ fill out stand out against the comments.  (the current impl doesn't look at the `required` flag because the data is not particularly correct in schemas...)
  - Fold subtrees of all comments so that you're not presented with a nightmare of 955(!) lines of YAML (for e.g. a Deployment)
  - If you start editing a value, that line all of its parent keys should be uncommented so that your change becomes part of the document

This may not cover everything but the basic algorithm is:
- Loop over all the fields of the input type

- If the path to the key is in the NEVER_ADD list, do not output anything, move on to next key.  These are things that are rarely/never useful for a human to be setting themselves.  They are complete paths starting from the root, i.e. "status" only matches at the top-level, not any field called status anywhere in the tree.

- If the key is not in ALWAYS_ADD, output it as a comment
- If it's in the ALWAYS_ADD list, output the key (uncommented).
  - All of the parent keys for this value should also become uncommented (even if they would have previously been comments).

- Recurse to produce the value:
  - If the value is a "simple" type, output it's type as the value (`foo: string`).  If the key isn't already commented by above, make the value a comment. (`#foo: string`, or `foo: #string`)
  - If it's an array, output one example entry for the value (by recursing on the schema for the value's type)
  - If it's a map, output one example entry (recurse)
  - If it's another resource type, output it (recurse)